agent/engine/dockerclient/dockerclientfactory.go

```diff
@@ -110,7 +120,15 @@ func (f *factory) getClient(version DockerVersion) (dockeriface.Client, error) {
 // findDockerVersions loops over all known API versions and finds which ones
 // are supported by the docker daemon on the host
 func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
-	clients := make(map[DockerVersion]dockeriface.Client)
+	// if the client version returns a MinAPIVersion and APIVersion, then use it to return
+	// all the Docker clients between MinAPIVersion and APIVersion
+	clients, err := findDockerVersionsfromMinMaxVersions(endpoint)
```

Comment:
Two concerns here. The first is that this is going to behave differently on windows. Windows only supports API version 1.24 or greater. If we put this check here, it will skip over the capability logic that ensures tasks are schedule-able to windows.

Reply:
Yeah, I agree. I missed this in my review. Good call!

Reply:
According to current logic for `FindSupportedAPIVersions`, if minimum version client `1.17` is not supported in Windows, the tasks should anyway not be scheduled on Windows right?

Reply:
No, not quite right.  We register capabilities for API versions below 1.24 on Windows because we still want tasks that use features in those APIs to work on Windows too.  Our capability logic is pretty naive right now: we map a set of features to the API version in which they were introduced and then use that as an indicator of the minimum version of Docker they'll run on.  Because Docker on Windows still supports (many of) the features that work on older versions of Docker, this was an inexpensive way to get that support without rebuilding our capability logic.

I agree that this is confusing and non-obvious.  I think long-term we probably want to revisit this logic and see if we can improve it, given that we now know different things from when we did when it was designed.  However, we haven't done that yet and we still need to register capabilities for older API versions on Windows.

Reply:
Got it. 
From reading today's code, in `findDockerVersions` for windows, we iterate through the `KnownAPIVersions` from 1.17. Wouldn't trying to create a client with version 1.17 or trying to ping, fail in Docker Windows? Eventually, those versions would not be added to the  list of clients that Docker supports and hence the capability for 1.17 would not be added? 
https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/dockerclient/dockerclientfactory.go#L115

Reply:
We don't actually want to create a client at 1.17 on Windows or even to attempt to ping it; we know that it'll fail (except on Docker 1.12 beta).  I'm talking strictly about capability registration here in that we still need to register those capabilities (and then override the incoming API version with the minimum of 1.24).

Reply:
But yes, I agree we should revisit this logic sometime and adapt it for windows.